bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/dto/thermostat/DateDeserializer.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.ecobee.internal.dto.thermostat;
+
+import java.lang.reflect.Type;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link DateDeserializer} is responsible for handling the UTC dates returned from
+ * the Ecobee API service.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@NonNullByDefault
+public class DateDeserializer implements JsonDeserializer<Date> {
+
+    @Override
+    public @Nullable Date deserialize(JsonElement element, Type arg1, JsonDeserializationContext arg2)
+            throws JsonParseException {
+        String date = element.getAsString();
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
+        try {
+            return formatter.parse(date);
+        } catch (ParseException e) {
+            return null;
+        }
+    }
+}
```

Comment:
Would you consider the deserializing directly to the intended target type, e.g. `ZonedDateType` instead? See for example:
https://github.com/openhab/openhab-addons/blob/d5dd3c9ea3a2ca609707d7e9b3b51bebf9a48432/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/deserialization/NADeserializer.java#L45-L57

Also, can you explain how this deserializer is used, since the DTO has changed from `Date` to `String`? I'm a little confused by this deserializer and the new version of `undefOrLocalDate` taking a String.

Reply:
> Would you consider the deserializing directly to the intended target type, e.g. ZonedDateType instead?

I would, but I'm not sure how to do it in this case.

The only changes in the DTO are for two date/time elements representing local time of the form "yyyy-MM-dd HH:mm:ss"

I left the Date elements intact for the remaining date/time elements representing UTC time also of the form "yyyy-MM-dd HH:mm:ss"

The problem I was having was that Ecobee uses the same date/time format (i.e. "yyyy-MM-dd HH:mm:ss") to represent local time and UTC time. Bad design on their part, since the only way to know if a date field is UTC or local time is to look at the Ecobee API documentation. 😢 

So I had to find a way to differentiate how the dates were handled. My solution was to use the Date deserializer for the dates that are in UTC time (the majority), which would force it to be UTC time by explicitly setting the timezone to UTC in the deserializer. In the case of the dates in local time (of which there are only 2), I chose to make them Strings, and handle them explicitly using the method `undefOrLocalDate`.

Hopefully this makes sense. If there's a better way to do it, let's discuss.

Reply:
Thanks for the explanation. Interesting, and not quite straight-forward, it seems. Without going more into the code right now, all I can say is that it would be nice if you could skip conversion to `Date` as intermediate type, since this is an old Java 8 type. Instead you could use `ZonedDateTime` when you need the time-zone or `Instant` when you don't. However, with your explanation, this seems to require some additional refactoring which is probably outside of the scope of this PR. So I will leave this comment and wait for your response, and I'd be happy to merge as is.

Reply:
Thanks.

It's not urgent to merge this, and I don't think there are that many places where Date is used. So, I'll take a look at changing it, so long as I can still accomplish what I need to (which I see no reason why I can't) and that it doesn't require too many additional changes.

Reply:
I forgot to mention `LocalDateTime`, which may also be of use. For example, DTO's could return either `LocalDateTime` or `Instant`, depending on source format. Applying time-zone from `TimeZoneProvider` could then be handled later and quite in the same fashion.

Reply:
Does GSON know how to deal with a DTO element defined as an `Instant`? From what I can see, I might need a custom deserializer if I want to use `Instant` in the DTOs for my UTC dates.

Reply:
No, that would require a custom deserializer.

Reply:
Yep, Instant and LocalDateTime both require custom deserializers. If you would please have a look at my changes. In the Instant deserializer, I had to first convert to LocalDateTime because I was getting exceptions when trying to parse straight to an Instant. I think that was because the format didn't conform what the parser required. This approach required a few more changes, but I also think it's cleaner.

Reply:
I agree, it's nice and clean - well done!